---
title: "Trait-based Comparative Transcriptomics Analysis"
author: "Joris van Steenbrugge"
date: "August 23, 2017"
output: html_document
---
```{r set-options, echo=FALSE, cache=FALSE}
options(width = 1400)
```
<style>
  .col2 {
    columns: 2 200px;         /* number of columns and width in pixels*/
    -webkit-columns: 2 200px; /* chrome, safari */
    -moz-columns: 2 200px;    /* firefox */
  }
  .col3 {
    columns: 3 100px;
    -webkit-columns: 3 100px;
    -moz-columns: 3 100px;
  }
</style>

```{r}
source('~/tools/tct/R/preprocess.R')
```
#Loading and Preprocessing Data


All of the preprocessing is handled by the `Pre_process_input` function.
This function consequtively performs the following tasks:

1. Data Loading
2. Normalization
3. Filtering


### Data loading
The data set should contain the following Elements:

* A column named **Gene**
* Multiple expression data columns that have at least the word **Sample** in it
* A column named **Annotation**
* A column named **Bin**

Additional columns may be present but are not used in the analysis.
Loading of data is done by assigning a file.path to the function's `filepath`
variable.
```{r}
file.path   <- "/home/joris/tools/samplematrix.csv"
RNAseq.data <- Pre_process_input(filepath = file.path,
                                 normalize.method = 'default',
                                 filter.method = 'MAD')

#Visualize 
knitr::kable(RNAseq.data$table[1:10, ])

```
### Normalization
The package supports different options for normalization. By default the `Pre_process_input` function
does not normalize the data. To indicate if normalization should be performed, the function parameter 
`normalize.var` may be used. The variable can either left out, set to 'default' or to a custom, user-defined, function.




